### PR TITLE
[Visual] Visual xaml and Visual registrations

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

- Add a type converter into core.design that's able to display available visuals
- Add a public method on Visual Type Converter in case people need to register there own names to an IVisual. This should help if there are visual collisions or they just want to call it something else
- Added registration of Visual types to converter during assembly scanning processes

image

### API Changes ###

Added:
 - VisualTypeConverter.Register(string visualKey, IVisual visual)
 

### Platforms Affected ### 

- Core/XAML (all platforms)

### Testing Procedure ###
- to test xaml changes you have to take the nuget and install it into a project then make sure you get Visual options when adding Visual to elements

### PR Checklist ###

- [x] Has automated tests 
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	{
+		static Dictionary<string, IVisual> _visualTypeMappings;
+		static Dictionary<string, IVisual> InitMappings()
+		{
+			var mappings = new Dictionary<string, IVisual>(StringComparer.OrdinalIgnoreCase);
+			_visualTypeMappings = mappings;
+			return mappings;
+		}
+
+		public static void Register(string visualKey, IVisual visual)
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: The idea here is allowing people get around naming collisions if they need to

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+			if(!_visualTypeMappings.ContainsKey(visual.FullName))
+			{
+				try
+				{
+					IVisual registeredVisual = (IVisual)Activator.CreateInstance(visual);
+					_visualTypeMappings[visual.Name] = registeredVisual;
+					_visualTypeMappings[visual.FullName] = registeredVisual;
+					_visualTypeMappings[$"{visual.Name}Visual"] = registeredVisual;
+					_visualTypeMappings[$"{visual.FullName}Visual"] = registeredVisual;
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: I added a few permutations of possible names people can use. If there are two Material visual libraries then people could use the full name of the type to resolve to what they want

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		internal static void Register(Type visual)
+		{
+			if (_visualTypeMappings == null)
+				InitMappings();
+
+			if(!_visualTypeMappings.ContainsKey(visual.FullName))
+			{
+				try
+				{
+					IVisual registeredVisual = (IVisual)Activator.CreateInstance(visual);
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: Not crazy about this. Need to see if there's a better way to do this. Just not sure yet how to pull out the IVisual type for the converter to use

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+					if (name.EndsWith("Visual", StringComparison.OrdinalIgnoreCase))
+					{
+						name = name.Substring(0, name.Length - 6);
+						fullName = fullName.Substring(0, fullName.Length - 6);
+					}
+
+					_visualTypeMappings[name] = registeredVisual;
+					_visualTypeMappings[fullName] = registeredVisual;
+					_visualTypeMappings[$"{name}Visual"] = registeredVisual;
+					_visualTypeMappings[$"{fullName}Visual"] = registeredVisual;
```
Comment (mattleibow) [inline Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: With the rename of the visual to XxxVisual, we probably have to check the reverse: 
if endsWith("Visual"), add (name-"Visual)

File: Xamarin.Forms.Core/Xamarin.Forms.Core.csproj [hunk 5]
```diff
@@ -22,17 +22,8 @@
...
-		Name="XFCorePostProcessor"
-		Inputs="$(IntermediateOutputPath)$(TargetFileName)"
-		Outputs="$(IntermediateOutputPath)XFCorePostProcessor.stamp">
-	    <Touch
-		    Files="$(IntermediateOutputPath)XFCorePostProcessor.stamp"
-		    AlwaysCreate="True" />
-	    <FixXFCoreAssembly
-		    Assembly="$(IntermediateOutputPath)$(TargetFileName)"
-		    ReferencePath="@(ReferencePath)" />
+	<Target Condition="$(DesignTimeBuild) != true AND $(BuildingProject) == true" AfterTargets="AfterCompile" Name="XFCorePostProcessor" Inputs="$(IntermediateOutputPath)$(TargetFileName)" Outputs="$(IntermediateOutputPath)XFCorePostProcessor.stamp">
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Xamarin.Forms.Core.csproj]: Dev16 auto changing project files

File: Xamarin.Forms.Core.Design/Xamarin.Forms.Core.Design.csproj [hunk 6]
```diff
@@ -50,6 +50,7 @@
     <Compile Include="..\Xamarin.Forms.Core\Properties\GlobalAssemblyInfo.cs">
       <Link>Properties\GlobalAssemblyInfo.cs</Link>
     </Compile>
+    <Compile Include="DerivedTypesConverter.cs" />
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core.Design/Xamarin.Forms.Core.Design.csproj]: consider changing the design csproj to SDK-style :)

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: we don't want registration by code. That prevents Xaml Compilation. If registration is required (i.e. you can't find all the Visuals through assembly reflection), consider registering using assembly attributes.

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+			if (value != null)
+			{
+				IVisual returnValue = null;
+				if (_visualTypeMappings.TryGetValue(value, out returnValue))
+					return returnValue;
+
+				return VisualMarker.Default;
+			}
+
+			throw new InvalidOperationException($"Cannot convert \"{value}\" into {typeof(IVisual)}");
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: you mean XamlParseException

File: Xamarin.Forms.Core/Xamarin.Forms.Core.csproj (hunk 5 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Xamarin.Forms.Core.csproj]: you know you can pick the cadges you want to include in your commits, while doing `git add` ? use `git add -i`, or `git add -p`

Comment (PureWeen) [reply Xamarin.Forms.Core/Xamarin.Forms.Core.csproj]: Yea :-/ but these are changes dev16 is going to make every time we open the SLN so it's easier to just get the project to a state of conforming

File: Xamarin.Forms.Core.Design/VisualTypes.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			if(Values == null)
+			{
+				var derivedNames = new List<string>();
+				var baseType = typeof(IVisual);
+
+				var typeElements = typeof(View).Assembly.ExportedTypes.Where(t => baseType.IsAssignableFrom(t));
+
+				foreach (var typeElement in typeElements)
+					if (typeElement.Name != baseType.Name)
```
Comment (mattleibow) [inline Xamarin.Forms.Core.Design/VisualTypes.cs]: Maybe move this if into the Where above:
`.Where(t => baseType.IsAssignableFrom(t) && t.Name != baseType.Name);`

Comment (PureWeen) [reply Xamarin.Forms.Core.Design/VisualTypes.cs]: 👍

File: Xamarin.Forms.Core.Design/VisualTypes.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
+﻿namespace Xamarin.Forms.Core.Design
+{
+	using System.Collections.Generic;
+	using System.Linq;
+	using System.ComponentModel;
+	using System;
+
+	public class VisualTypes : TypeConverter
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core.Design/VisualTypes.cs]: type converters are usually named FooTypeConverter

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core.Design/VisualTypes.cs]: oh, it's in Design, I don't know about the naming there

File: Xamarin.Forms.Core/ReflectionExtensions.cs [hunk 10]
```diff
@@ -45,6 +45,31 @@ public static PropertyInfo GetProperty(this Type type, string name)
 			return null;
 		}
 
+		internal static object[] GetCustomAttributesSafe(this Assembly assembly,  Type attrType)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/ReflectionExtensions.cs]: we should move, some time later, all this into the registrar. but fine

File: Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs [hunk 11]
```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Xamarin.Forms.Internals;
+using Xamarin.Forms.Xaml;
+
+namespace Xamarin.Forms
+{
+	[Xaml.TypeConversion(typeof(IVisual))]
+	public class VisualTypeConverter : TypeConverter
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Visuals/VisualTypeConverter.cs]: add the attribute to accept null service provider

**Pull request merged**